displace_node_list_by_side_list option

roystgnr opened the pull request:
This should be one step toward resolving #30385

## Reason
Currently, when `construct_node_list_from_side_list` is true (as is default, and as is currently necessary for much of MOOSE), MOOSE will happily merge side list nodes into any existing unrelated nodeset.  This is bad.

## Design
Add option, `displace_node_list_by_side_list`, to renumber any existing nodeset that would otherwise be merged into by a sideset but does not share the sideset's name.

## Impact
Currently no impact on users who don't manually enable the option.

This option really should default to `true`, but doing that breaks about 50 exodiffs in test/ and modules/, so I'll make it a separate PR after I've double-checked each diff; just getting the option at all has become very urgent.

roystgnr committed:
Fix construct_node_list_from_side_list doc string

Using a nodeset as a cache to a sideset was a mediocre idea.  Assuming
that no distinct nodeset exists with the same id as a sideset was a bad
idea.  Silently merging the two when a distinct node does exist was
a horrible idea.  The phrase "good idea" belongs nowhere near this
option.

roystgnr committed:
Add "displace_node_list_by_side_list" option

Setting this to "false" for now while I'm debugging, but really this
ought to be the default ASAP.

We'll only displace a node list when constructing it from a side list.
If a user is sophisticated enough not to spam node lists from side lists
then they probably don't want us messing with their node lists at all,
even after we make displace the default behavior.

Refs #31895

roystgnr committed:
Regression test for nodeset displacement

roystgnr committed:
Default displace_node_list_by_side_list=true

I can't imagine there's anyone who really wants the previous behavior.

Now let's see which regression gold files really want it.

GiudGiud commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -3020,8 +3028,38 @@ MooseMesh::getBoundaryIDs() const
   if (_construct_node_list_from_side_list)
-    getMesh().get_boundary_info().build_node_list_from_side_list();
+  {
+    if (_displace_node_list_by_side_list)
```
need to error if displace node list is set when construct_node_list is not since it s ignored

roystgnr replied:
Actually - this shouldn't apply if we can default displace_ to true, should it?  If someone's disabling construct_ we don't want their input file to start erroring after this PR.

roystgnr replied:
I'll do that in a subsequent PR; in this PR I'm having to leave it disabled by default (Bison fixes are in a PR now, but that'll take time and SAM'll take longer) so I added the error.

roystgnr committed:
Ensure new nodesets get names of their sidesets

This is needed to make the nodeset displacement operation idempotent,
but also: why were we not doing this already?

GiudGiud commented on test/tests/mesh/node_list_from_side_list/tests:
```diff
@@ -1,11 +1,20 @@
+    exodiff = 'displace_node_list_in.e'
+    recover = false
+    issues = '#30385'
+    requirement = 'The system shall allow displacing rather than merging into a nodeset whose id conflicts with a sideset'
```
```suggestion
    requirement = 'The system shall be able to create a sufficient offset in the nodeset ids rather than merge converted sidesets into a nodeset whose id conflicts with a sideset'
```

roystgnr commented:
Ooh, actually, one fix gets rid of at least all the test/ failures for me.  I'll see if modules/ and CI agree.

roystgnr committed:
Apply suggestion from @GiudGiud

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

GiudGiud commented:
```
[Apply suggestion from] @GiudGiud
```
if you tag me in a commit I will get a zillion emails unfortunately

roystgnr committed:
Clearer displace_node_list_by_side_list doc string

GiudGiud reviewed (approved):
your call on adding the error, with the parameter text and the default we are mostly good

GiudGiud commented on test/tests/mesh/node_list_from_side_list/displace_node_list.i:
```diff
@@ -0,0 +1,7 @@
+[Mesh]
+  [gen]
+    type = FileMeshGenerator
+    file = bcid_conflict.xda
```
looks like you forgot to check in this file

use -f maybe

roystgnr replied:
Of course, there's a .gitignore line ignoring .xda.  Why would we ever want to use a file format that supports more features in 30% (or 10% if you gz it) of the size?

(I lash out because I'm embarrassed; thanks for catching this)

roystgnr committed:
Add input file for displace_node_list test

I missed this before; .xda is in our .gitignore...

roystgnr committed:
Disable displace_node_list_by_side_list for now

It seems to be interfering with a regression test in SAM and with about
50 regression tests in Bison.

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3379193/), step Results summary on d2d2e14 wanted to post the following:

## Framework test summary

Compared against 7edd10c in job [civet.inl.gov/job/3374418](https://civet.inl.gov/job/3374418).

### Added tests

| Test                                               | Time (s)   |
|----------------------------------------------------|------------|
| `mesh/node_list_from_side_list.displace_node_list` | 0.81       |
## Modules test summary

Compared against 7edd10c in job [civet.inl.gov/job/3374418](https://civet.inl.gov/job/3374418).

No change

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3379159/), step Generate coverage on d2d2e14 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>7edd10</th>
      <th colspan=4>#31895 d2d2e1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.97%</td>
      <td>85.97%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>124954</td>
      <td>124974</td>
      <td>+20</td>
      <td>23</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>20387</td>
      <td>20388</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31895/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31895/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31895/coverage/xfem)

</details>



This comment will be updated on new commits.

roystgnr committed:
Add missing design entry to new test

roystgnr committed:
Avoid wrapping new nodeset ids

Hopefully this will fix an error in a SAM regression test
